testing/tests/DevExpress.ui.widgets/treeViewParts/keyboardNavigation.js

```diff
@@ -78,11 +81,206 @@ QUnit.test('down arrow move focus to the next element for virtual mode & slow DS
 
     $firstItem.trigger('dxpointerdown');
 
-    assert.ok($firstNode.hasClass('dx-state-focused'), 'item was focused ');
+    assert.ok($firstNode.hasClass(FOCUSED_STATE_CLASS), 'item was focused ');
 
     keyboard.keyDown('down');
 
-    assert.ok($secondNode.hasClass('dx-state-focused'), 'next item was focused after \'down\' was pressed');
+    assert.ok($secondNode.hasClass(FOCUSED_STATE_CLASS), 'next item was focused after \'down\' was pressed');
+}),
+
+QUnit.test('focus should change from selectAll item to first item on down key pressed', function(assert) {
+    if(!isDesktopDevice()) {
+        assert.ok(true, 'only on desktops');
+        return;
+    }
+
+    const $treeView = initTree({
+        showCheckBoxesMode: 'selectAll',
+        items: [ { id: '1' } ]
+    });
+
+    $treeView.focusin();
+    keyboardMock($treeView).keyDown('down');
+
+    const $firstItem = $(`.${internals.NODE_CLASS}`).eq(0);
+    assert.ok($firstItem.hasClass(FOCUSED_STATE_CLASS));
+}),
+
+QUnit.test('focus should change from first item to selectAll item on up key pressed', function(assert) {
+    if(!isDesktopDevice()) {
+        assert.ok(true, 'only on desktops');
+        return;
+    }
+
+    const $treeView = initTree({
+        showCheckBoxesMode: 'selectAll',
+        items: [ { id: '1' } ]
+    });
+
+    $treeView.focusin();
+    const $firstItem = $(`.${internals.NODE_CLASS}`).eq(0);
+    $firstItem.trigger('dxpointerdown');
+    keyboardMock($treeView).keyDown('up');
+
+    const $selectAllItem = $(`.${internals.SELECT_ALL_ITEM_CLASS}`);
+    assert.ok($selectAllItem.hasClass(FOCUSED_STATE_CLASS));
+}),
+
+QUnit.test('focus should change from selectAll item to last item on up key pressed', function(assert) {
+    if(!isDesktopDevice()) {
+        assert.ok(true, 'only on desktops');
+        return;
+    }
+
+    const $treeView = initTree({
+        showCheckBoxesMode: 'selectAll',
+        items: [ { id: '1' }, { id: '2' } ]
+    });
+
+    $treeView.focusin();
+    keyboardMock($treeView).keyDown('up');
+
+    const $lastNode = $(`.${internals.NODE_CLASS}`).eq(1);
+    assert.ok($lastNode.hasClass(FOCUSED_STATE_CLASS));
+}),
+
+QUnit.test('focus should change from last item to selectAll item on down key pressed', function(assert) {
+    if(!isDesktopDevice()) {
+        assert.ok(true, 'only on desktops');
+        return;
+    }
+
+    const $treeView = initTree({
+        showCheckBoxesMode: 'selectAll',
+        items: [ { id: '1' }, { id: '2' } ]
+    });
+
+    $treeView.focusin();
+    const $lastNode = $(`.${internals.NODE_CLASS}`).eq(1);
+    $lastNode.trigger('dxpointerdown');
+    keyboardMock($treeView).keyDown('down');
+
+    const $selectAllItem = $(`.${internals.SELECT_ALL_ITEM_CLASS}`);
+    assert.ok($selectAllItem.hasClass(FOCUSED_STATE_CLASS));
+}),
+
+QUnit.test('selectAll item should be checked when navigating up from first item with shift key', function(assert) {
```

Comment:
I think it shouldn't. If we navigate up several times and accidentally came to selectAll, all items will be selected